Use `poll(2)`/`ppoll(2)` instead of `select(3P)`

dinosaure opened the pull request:
This pull request proposes using `poll`/`ppoll` (from [iomux](https://github.com/ocaml-multicore/ocaml-iomux)) instead of `select()` so that `miou.unix` can function correctly on systems such as FreeBSD, which fail when the value of a file descriptor exceeds 1023. In the case of `poll`/`ppoll`, the limit of 1024 active file descriptors remains, but their values can exceed 1023 and they can be passed to `poll`/`ppoll`.

This PR is a draft. The `Bitv` module is not tested and `Bitv.max` does not work (but is not used).

Another more significant problem is that we have noticed a performance regression with [`httpcats`](https://github.com/robur-coop/httpcats). With benchmarks and metrics, a notable difference between `select(3P)` and `poll(2)`/`ppoll(2)` relates to the garbage collector:
- in the case of `select(3P)`, we need to allocate the lists of file descriptors that we would like to give to `Unix.select`. It seems that this allocation triggers the GC before calling `Unix.select`. We would need to look into this in detail, but this pre-call to the GC before `Unix.select` (and specifically `caml_enter_blocking_section()`) saves us from a call to the GC just before our actual call to `select(3P)`.
- In the case of `poll(2)`/`ppoll(2)`, no additional allocation is necessary to use these functions. `caml_enter_blocking_section()`, which also exists before our call to `poll(2)`/`ppoll(2)`, seems more likely to call the GC. A simple benchmark with `httpcats` suggests that if we add a `Gc.minor` before, the application can handle more HTTP requests (but we note a large regression in latency).

The regression between `select(3P)` and this PR is measured in terms of the number of requests per second and the latency of our application:
- With `select(3P)`: 312k req/s and a latency of 0.82ms
- With this PR: 223k req/s and a latency of 1.15ms

It therefore seems that `poll(2)`/`ppoll(2)` gives the GC more opportunity to clean up the OCaml heap, which (macro-wise) degrades the speed of our application.

I would like to give special thanks to @haesbaert, who I think will be very interested in these results.

dinosaure committed:
Add iomux/poll/ppoll support on miou.unix

dinosaure committed:
Fix the compilation of bitv.c (about C casting)

dinosaure committed:
Apply formatter on dune files

dinosaure commented:
For greater precision, here is the FlameGraph with `select(3P)`:
![perf select](https://github.com/user-attachments/assets/29cdc233-676a-41ce-a8b6-c348fa142780)

And here is the FrameGraph with `ppoll(2)`:
![perf iomux](https://github.com/user-attachments/assets/07dab6e4-26ac-49d8-81be-a3ebaeea2679)

dinosaure committed:
Fix Bitv.max

dinosaure committed:
Add licenses and documentations

hannesm commented:
looks fine, I'm curious whether we should get another bitv implementation since it otherwise will impose the lgpl (or gpl?) on the package!? Doesn't look like a lot of work to redo it (or ask Jean-Christophe whether we can have it as MIT/ISC licensed piece).

dinosaure committed:
Cap the number of file-descrs

dinosaure commented:
To summarize what is missing:
- [ ] we should improve the documentation and we should not mention `select(3P)`
- [ ] the API does not change, but we probably can export `Miou_poll`, I'm not fully decided about that
- [ ] we must have tests about `Miou_bitv` (add a fuzzer?). Even if the implementation seems right, we must be sure that it's correct.

dinosaure commented:
> The way I see it, if we wanted to do that, we would expose each backend-specific module (Miou_poll, Miou_kqueue, etc.) and instantiate Miou_unix's common interface with one of those. That way, if a user wants their application/library to be generic they can use Miou_unix, and if they want platform-specific features they can use the appropriate backend module.

It's indeed really interesting. This question can be quite difficult in the sense that there may be two mechanisms in OCaml for dispatching according to an implementation:
1) we can dispatch the loop dynamically, which would allow a Miou application to run on any system even if it was compiled on Linux (for example)
2) You can dispatch the loop statically, thereby enforcing the fact that the application compiled with Miou is strictly dependent on the system on which it was compiled.

I would be more inclined towards static dispatch as long as it is easy to cross-compile (particularly with Docker) an application on a specific system that you do not own.

This is in line with the approach you suggest with a common `Miou_unix` interface, the implementation of which would depend on the system we are on and would be chosen at compile time.

However, this raises another question about the specific features that a system may have and how to make them available when `Miou_unix` would be common to all platforms.

This is where I might tend (but we need to look at the details) to propose `Miou_kqueue`, `Miou_poll` and `Miou_io_uring` explicitly rather than offering a common `Miou_unix`. The question is very interesting, however, and I am sure that others may also have an opinion on the subject. With this in mind, we would like to raise an issue in order to clearly identify the problem and how to solve it.

dinosaure committed:
Change the license of miou_bitv.ml{,i} to MIT (from LGPL-2.1-or-later)

Jean-Christophe Filliatre authorized us to change the license of miou_bitv.ml{,i}

dinosaure committed:
Fix the support of pkg/pkg.ml and change the layout to compile miou

dinosaure committed:
Apply formatter on dune files

dinosaure committed:
Improve myocamlbuild.ml and fix B0.ml (with b0.0.0.6)

dinosaure committed:
Complete the test about bitv

pukkamustard commented:
> I'm currently working on a [`kqueue(2)`](https://man.openbsd.org/kqueue.2) backend for `Miou_unix` (which is not public / ready yet, it will have its own PR when the time comes

Have you been able to progress on this? I'm interested in a kqueue backend.

dinosaure commented:
> Have you been able to progress on this? I'm interested in a kqueue backend.

I think some people (like @hannesm) will be interesting about that. The current design of `miou` can help, during compilation, to choose between `poll`/`ppoll`/`select` and `kqueue` depending on the system on which the project is being compiled. This shouldn’t be too difficult, and I’d be happy to help and merge such a proposal (unfortunately, I’m not a FreeBSD user).

omegametabroccolo commented:
@pukkamustard Hi, unfortunately I'm no longer interested in this. I may still have some half-working code lying around somewhere, if I find it I'll make sure to share it, but no promises :(
